blockprod/src/detail/mod.rs

```diff
@@ -47,158 +85,566 @@ impl BlockProduction {
+    ) -> Result<(ConsensusData, GenBlockIndex), BlockProductionError> {
+        let consensus_data = self
             .chainstate_handle
-            .call(|this| {
-                if let Ok(current_tip_id) = this.get_best_block_id() {
-                    if let Ok(Some(current_tip_height)) =
-                        this.get_block_height_in_main_chain(&current_tip_id)
-                    {
-                        return Some((current_tip_id, current_tip_height));
-                    }
+            .call({
+                let chain_config = Arc::clone(&self.chain_config);
+
+                move |this| {
+                    let best_block_index = this
+                        .get_best_block_index()
+                        .expect("Best block index retrieval failed in block production");
+
+                    let get_ancestor = |block_index: &BlockIndex, ancestor_height: BlockHeight| {
+                        this.get_ancestor(
+                            &block_index.clone().into_gen_block_index(),
+                            ancestor_height,
+                        )
+                        .map_err(|_| {
+                            PropertyQueryError::GetAncestorError(
+                                GetAncestorError::InvalidAncestorHeight {
+                                    block_height: block_index.block_height(),
+                                    ancestor_height,
+                                },
+                            )
+                        })
+                    };
+
+                    let consensus_data = consensus::generate_consensus_data(
+                        &chain_config,
+                        &best_block_index,
+                        block_timestamp,
+                        best_block_index.block_height().next_height(),
+                        get_ancestor,
+                    );
+                    consensus_data.map(|cons_data| (cons_data, best_block_index))
                 }
-
-                None
             })
             .await?
-            .ok_or(BlockProductionError::FailedToConstructBlock(
-                BlockCreationError::CurrentTipRetrievalError,
-            ))?;
-
-        let (_tx, dummy_rx) = crossbeam_channel::unbounded();
-
-        let block_maker = BlockMaker::new(
-            self.chain_config.clone(),
-            self.chainstate_handle.clone(),
-            self.mempool_handle.clone(),
-            self.time_getter.clone(),
-            reward_destination,
-            current_tip_id,
+            .map_err(BlockProductionError::FailedConsensusInitialization)?;
+
+        Ok(consensus_data)
+    }
+
+    fn solve_block(
+        chain_config: Arc<ChainConfig>,
+        mut block_header: BlockHeader,
+        current_tip_height: BlockHeight,
+        stop_flag: Arc<AtomicBool>,
+    ) -> Result<BlockHeader, BlockProductionError> {
+        // TODO: use a separate executor for this loop to avoid starving tokio tasks
```

Comment:
You can use https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html. The task will be allocated to a thread that does not starve other tasks.

Reply:
We're using SlavePool. This is an outdated comment.